### PR TITLE
Use map to store boundary conditions instead of vectors

### PR DESCRIPTION
**Pull request by blaisb**
### Description

The way we stored boundary condition was always poorly designed. We used a vector with ids that would correspond to the deal.II IDs, but we also had the lethe-ids which were contiguous. This was very confusing (which id is supposed to be what?) and it also meant we always had to search through the vector of ids.

I refactored this entirely to instead use a map where the key is the boundary id and the values is the type (or the function, or etc). This way, we never have to look up what is "our id", "our id" is always the ID we put in the mesh generator for the boundary condition.

This required some extensive refactoring, but in the end i find that the final design is much much cleaner.

### Testing

This does not include new test. I had to remake ONE restart file for ONE test. That is the fluidized bed case. The issue is that the order in which we apply the boundary conditions changed in this case and so that actually change the reslts because the corner nodes for the velocity are a bit difference. I used this opportunity to make the test a bit better and more reproducible.

### Documentation

Nothing to document

### Miscellaneous (will be removed when merged)

While people are reviewing this PR I willuse this opportunity to add more assertions there and there. This new mechanism offers us plenty of opportunity to actually double check our code adequately.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [ ] Lethe documentation is up to date
- [x] The branch is rebased onto master
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent
- [x] If parameters are modified, the tests and the documentation of examples are up to date
- [ ] Changelog (CHANGELOG.md) is up to date if the refactor affects the user experience or the codebase

Pull request related list:
- [x] No other PR is open related to this refactoring
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If any future works is planed, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: applications_tests/lethe-fluid-block/cylinder_carreau_gd.prm [hunk 1]
```diff
@@ -94,11 +94,13 @@ end
 #---------------------------------------------------
 
 subsection boundary conditions
-  set number = 3
+  set number = 4
   subsection bc 0
+    set   id = 0
```
Comment (lpsaavedra) [inline applications_tests/lethe-fluid-block/cylinder_carreau_gd.prm:99]: I don't think this is correctly indented. I see the same mistake in several parameter files. 
```suggestion
    set   id = 0
```

**Review by lpsaavedra — COMMENTED** [review]
As a first comment, I suggest applying the prmindent script to all files. There are more than 80 scripts that are not correctly indented :)

File: examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm
Comment (lpsaavedra) [inline examples/incompressible-flow/3d-turbulent-taylor-couette/tc-matrix-free.prm:1]: The id's for the first two boundary conditions are missing

File: include/core/boundary_conditions.h [hunk 2]
```diff
@@ -61,29 +92,27 @@ namespace BoundaryConditions
...
 
     // Number of boundary conditions
-    unsigned int size;
+    unsigned int number_of_boundary_conditions;
 
     /// indicator for transient BCs
     bool time_dependent;
 
 
     // Periodic boundary condition matching
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:112]:
```suggestion
    /// Map containing the boundary id and its corresponding periodic boundary condition match
```

File: include/core/boundary_conditions.h [hunk 3]
```diff
@@ -61,29 +92,27 @@ namespace BoundaryConditions
...
-    std::vector<unsigned int> periodic_id;
-    std::vector<unsigned int> periodic_direction;
+    std::map<types::boundary_id, types::boundary_id> periodic_neighbor_id;
+
+    std::map<types::boundary_id, unsigned int> periodic_direction;
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:115]:
```suggestion
    /// Map containing the boundary id and its periodic direction
    std::map<types::boundary_id, unsigned int> periodic_direction;
```

File: include/core/boundary_conditions.h [hunk 4]
```diff
@@ -101,6 +130,9 @@ namespace BoundaryConditions
     Functions::ParsedFunction<dim> v;
     Functions::ParsedFunction<dim> w;
 
+    // Pressure
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:133]:
```suggestion
    /// Pressure
```

File: include/core/boundary_conditions.h [hunk 5]
```diff
@@ -101,6 +130,9 @@ namespace BoundaryConditions
...
+    Functions::ParsedFunction<dim> p;
+
     // Point for the center of rotation
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:136]:
```suggestion
    /// Point for the center of rotation
```

File: include/core/boundary_conditions.h [hunk 6]
```diff
@@ -61,29 +92,27 @@ namespace BoundaryConditions
...
     // dirichlet BCs
-    std::vector<double> boundary_layer_thickness;
+    std::map<types::boundary_id, double> boundary_layer_thickness;
 
     // Number of boundary conditions
-    unsigned int size;
+    unsigned int number_of_boundary_conditions;
 
     /// indicator for transient BCs
     bool time_dependent;
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:111]:
```suggestion
```

File: include/core/boundary_conditions.h [hunk 7]
```diff
@@ -144,14 +160,15 @@ namespace BoundaryConditions
   class NSBoundaryConditions : public BoundaryConditions<dim>
   {
   public:
-    // Functions for (u,v,w) for all boundaries
-    NSBoundaryFunctions<dim>         *bcFunctions;
-    NSPressureBoundaryFunctions<dim> *bcPressureFunction;
+    // Functions for (u,v,w,p) for all boundaries
```
Comment (lpsaavedra) [inline include/core/boundary_conditions.h:163]:
```suggestion
    /// Functions for (u, v, w, p) for all boundaries
```

File: include/core/utilities.h [hunk 8]
```diff
@@ -648,5 +648,43 @@ get_penalty_factor(const unsigned int fe_degree,
...
+  return keys;
+}
+
+/**
+ * @brief Extract the values from map object
+ *
+ * @tparam key Key of the map
+ * @tparam val Value of the map. This template parameter is not used
+ * @param map Map object from which the keys are extracted
+ * @return values Vector of values from the map
```
Comment (lpsaavedra) [inline include/core/utilities.h:676]:
```suggestion
 * @tparam key Key of the map. This template parameter is not used.
 * @tparam val Value of the map. 
 * @param map Map object from which the keys are extracted
 * @return values Vector of values from the map
```

File: include/solvers/navier_stokes_base.h [hunk 9]
```diff
@@ -42,6 +43,14 @@
...
 using namespace dealii;
 
+
+
+DeclException1(
+  FluidDynamicsBoundaryConditionMissing,
+  types::boundary_id,
+  << "The boundary id: " << arg1
+  << " is defined in the triangulation, but not as a boundary condition for the fluid dynamics physics has been defined for it. Lethe does not assign a    default boundary condition to boundary ids. Every boundary id defined    within the triangulation must have a corresponding boundary condition    defined in the input file.");
+
```
Comment (lpsaavedra) [inline include/solvers/navier_stokes_base.h:53]:
```suggestion
  << "The boundary id: " << arg1
  << " is defined in the triangulation, but no boundary condition has been set in the fluid dynamics physics section. Lethe does not assign a  default boundary condition to boundary ids. Every boundary id defined within the triangulation must have a corresponding boundary condition defined in the input file.");

```

File: include/solvers/tracer.h [hunk 10]
```diff
@@ -35,6 +35,8 @@
 #include <deal.II/lac/trilinos_sparse_matrix.h>
 #include <deal.II/lac/trilinos_vector.h>
 
+#include <map>
+
```
Comment (lpsaavedra) [inline include/solvers/tracer.h:39]:
```suggestion
```

File: source/solvers/tracer.cc [hunk 11]
```diff
@@ -31,6 +31,8 @@
 #include <deal.II/numerics/error_estimator.h>
 #include <deal.II/numerics/vector_tools.h>
 
+#include <map>
+
```
Comment (lpsaavedra) [inline source/solvers/tracer.cc:35]:
```suggestion
```

File: source/solvers/tracer.cc [hunk 12]
```diff
@@ -429,10 +428,7 @@ Tracer<dim>::assemble_system_rhs_dg()
       // Identify which boundary condition corresponds to the boundary id. If
       // this boundary condition is not identified, then exit the simulation
       // instead of assuming an outlet.
```
Comment (lpsaavedra) [inline source/solvers/tracer.cc:430]:
```suggestion
      // Identify which boundary condition corresponds to the boundary id. 
```

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
Great work, so many files changed, but I think this is a change that needed to be done.  This is my first batch of comments.

File: applications_tests/lethe-fluid-block/cylinder_carreau_gd.prm (hunk 1 above)
Comment (blaisb) [reply applications_tests/lethe-fluid-block/cylinder_carreau_gd.prm:99]: I just reapplied prmindent to all files :)

File: include/core/boundary_conditions.h [hunk 13]
```diff
@@ -449,24 +517,36 @@ namespace BoundaryConditions
...
     declare_parameters(ParameterHandler  &prm,
                        const unsigned int number_of_boundary_conditions);
     void
-    parse_boundary(ParameterHandler &prm, const unsigned int i_bc);
+    parse_boundary(ParameterHandler &prm);
     void
     parse_parameters(ParameterHandler &prm);
 
+
+    // TODO REMOVE THIS FLAG IT IS NOT USEFUL
```
Comment (oguevremont) [inline include/core/boundary_conditions.h:549]: Did you want to remove this flag now or in another PR?

**Review by oguevremont — COMMENTED** [review]
Very nice refactor! Only minor comments and suggestions

File: include/core/boundary_conditions.h [hunk 14]
```diff
@@ -1110,24 +1197,35 @@ namespace BoundaryConditions
...
   template <int dim>
   void
-  VOFBoundaryConditions<dim>::parse_boundary(ParameterHandler  &prm,
-                                             const unsigned int i_bc)
+  VOFBoundaryConditions<dim>::parse_boundary(ParameterHandler &prm)
   {
+    Assert(
+      prm.get_integer("id") >= 0,
+      ExcMessage(
+        "An invalid boundary id has been given for a VOF boundary condition."));
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:1208]: It's not really used to check if the boundary ID has not been set. It's really to ensure that the boundary IDs are positive

File: include/core/boundary_conditions.h [hunk 15]
```diff
@@ -944,32 +1024,42 @@ namespace BoundaryConditions
...
   template <int dim>
   void
-  CahnHilliardBoundaryConditions<dim>::parse_boundary(ParameterHandler  &prm,
-                                                      const unsigned int i_bc)
+  CahnHilliardBoundaryConditions<dim>::parse_boundary(ParameterHandler &prm)
   {
+    Assert(
+      prm.get_integer("id") >= 0,
+      ExcMessage(
+        "An invalid boundary id has been given for a Cahn Hilliard boundary condition."));
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:1035]: It's not really used to check if the boundary ID has not been set. It's really to ensure that the boundary IDs are positive

File: include/core/boundary_conditions.h [hunk 16]
```diff
@@ -811,15 +893,14 @@ namespace BoundaryConditions
   {
     prm.enter_subsection("boundary conditions tracer");
     {
-      this->size           = prm.get_integer("number");
-      this->time_dependent = prm.get_bool("time dependent");
-      this->type.resize(this->size);
+      this->number_of_boundary_conditions = prm.get_integer("number");
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:899]: Yup that's a good point, but I think this will be a very extensive further refactor. I can open an issue and refactor that in a next PR

File: include/core/boundary_conditions.h [hunk 17]
```diff
@@ -768,34 +842,42 @@ namespace BoundaryConditions
...
   template <int dim>
   void
-  TracerBoundaryConditions<dim>::parse_boundary(ParameterHandler &prm,
-                                                unsigned int      i_bc)
+  TracerBoundaryConditions<dim>::parse_boundary(ParameterHandler &prm)
   {
+    Assert(
+      prm.get_integer("id") >= 0,
+      ExcMessage(
+        "An invalid boundary id has been given for a tracer boundary condition."));
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:857]: It's not really used to check if the boundary ID has not been set. It's really to ensure that the boundary IDs are positive

File: include/core/boundary_conditions.h [hunk 18]
```diff
@@ -584,21 +651,30 @@ namespace BoundaryConditions
...
   template <int dim>
   void
-  HTBoundaryConditions<dim>::parse_boundary(ParameterHandler  &prm,
-                                            const unsigned int i_bc)
+  HTBoundaryConditions<dim>::parse_boundary(ParameterHandler &prm)
   {
+    Assert(
+      prm.get_integer("id") >= 0,
+      ExcMessage(
+        "An invalid boundary id has been given for a heat transfer boundary condition."));
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/core/boundary_conditions.h:662]: It's not really used to check if the boundary ID has not been set. It's really to ensure that the boundary IDs are positive

File: include/solvers/navier_stokes_base.h [hunk 19]
```diff
@@ -812,6 +822,29 @@ class NavierStokesBase : public PhysicsSolver<VectorType>
 
   // Member variables
 protected:
+  /**
+   * @brief Verify consistency of the input parameters for boundary
+   * conditions to ensure that for every boundary condition within the
+   * triangulation, a boundary condition has been specified in the input file.
+   */
+  void
+  verify_consistency_of_boundary_conditions()
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/navier_stokes_base.h:831]: Yeah that's a good point. The issue is that the function would have to take two arguments (the type of exceptions and the boundary objects). In the end, it would not be much leaner this way. This can be refactored afterwards though

File: include/core/boundary_conditions.h (hunk 13 above)
Comment (blaisb) [reply include/core/boundary_conditions.h:549]: Maybe in another PR. This PR is already kinda bloated in my opinion..

File: source/solvers/navier_stokes_assemblers.cc [hunk 20]
```diff
@@ -2386,108 +2320,109 @@ PartialSlipDirichletBoundaryCondition<dim>::assemble_rhs(
     1. / std::pow(scratch_data.cell_size, fe.degree + 1);
   auto &local_rhs = copy_data.local_rhs;
   // Loop over the BCs
-  for (unsigned int i_bc = 0; i_bc < this->boundary_conditions.size; ++i_bc)
+  {
```
[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/navier_stokes_assemblers.cc:2323]: Nope they are not. Good catch man

**Review by lpsaavedra — APPROVED** [review]

**Comment by blaisb** [comment]
Thank you all for the hard work. I will recheck the examples tomorrow morning and add a CHANGELOG entry :)!
